### ISSUE TITLE
Password hashing

### ISSUE THREAD

**Issue by TuxCoder** [gogits/gogs#1734]
Because of one of the last updates crashes my ldap config, I took a look how the local passwords are hashed.

I'm wondered why is PBKDF2 used as hashing algorithm. Haven't seen that yet.
Wikipedia says to PBKDF2 it is normally used for generating a key for a symmetric encryption like AES. https://en.wikipedia.org/wiki/PBKDF2
It looks not bad, but I recommend to use something like bcrypt.
https://godoc.org/golang.org/x/crypto/bcrypt

Reasons:
* it's securer (much higher cpu power is needed not good to calculate on a fpga)
* easier to update the hash algorithm to a higher difficult (difficult is saved in the hash)
* easier to change to an alternative because it uses a default save format with the algorithm type in the hash
example:`$2a$10$uYjSGJTU21mQ/RNhm9pRQuooOdqKCZqvHr8kf6biUiNlQbP.MdQIa10`

The salt is included into the hash so its easier to store.

Problems:
Changing the hashing algorithm is a bit a mess because we have to check is the saved password the old or the new hash. Probably do this over the salt. If salt == null new else old

Benchmark with cudaHashcat:
(code 10000) PBKDF2-SHA256 20000 loops: 7831H/s
(code 3200) bcrypt difficult 32 (max):  655 (think this is a false result)
(code 3200) bcrypt difficult 10 try to crack simple password: ~21H/s
I'm not sure how accurate the results with hashcate are because of the high different.

Benchmark implemented in go (probably there are faster ones):
bcrypt 1 core difficult 10 : ~10H/s
current hash 1 core: ~21H/s
brcypt 1 core difficult 20: ~0.01H/s == 101  s/H 

Has anyone else a better knowledge about that? How secure is PBKDF2 for password hashing?

ps. thx for this nice project

**Comment by Unknwon** [comment]
Thanks your feedback!

Initial reason for using it: https://github.com/gogits/gogs/issues/42

Others I can't... :sweat_smile:

**Comment by Eriner** [comment]
PBKDF2-SHA256 is a secure hashing algorithm. Crackstation has a good article on password hashing (and why not to make your own). https://crackstation.net/hashing-security.htm

**Comment by Unknwon** [comment]
@Eriner thanks your info!

**Comment by Eriner** [comment]
To be clear, though, bcrypt *is* a 'more secure' algorithm. There are cases where PBKDF2 has been implemented in FPGAs. PBKDF2 can be GPU accelerated, whereas bcrypt can not.

The real difference is logic-hard vs memory-hard.
Logic-hard is prime factoring, and memory-hard is lookup tables.

Memory-hard algorithms have not been implemented on acceleration platforms (too expensive), whereas logic-hard algorithms have. The risk is that memory-hard algorithms haven't been around long (comparatively), so no one is completely sure what potential attacks may exist against them.

I personally would argue **bcrypt > PBKDF2-SHA256**. 

(NIST recommends PBKDF2)[http://csrc.nist.gov/publications/nistpubs/800-132/nist-sp800-132.pdf], but take that for a grain of salt.

**Comment by Eriner** [comment]
It might be worth investigating replacing PBKDF2 with bcrypt for the next major milestone. 
https://godoc.org/golang.org/x/crypto/bcrypt

It's not that PBKDF2 is bad, just that bcrypt is better. 

Just to add to the information above, scrypt is ~4000 times slower than bcrypt (which is good), and uses more RAM for its jump-table. The worry is that scrypt has only been around for 4 years or so, and while no problems have yet been found in scrypt, there have been no reported problems with bcrypt either.

**Comment by Eriner** [comment]
@TuxCoder @Unknwon implementation of the switch should not be hard, just have it prompt the user for a password change on first login after bcrypt is implemented. Once most/all of the users have 'changed' (re-hashed with bcrypt) their passwords, it would be safe to remove the old PBKDF2 code. At that point, remaining users can reset their password the old-fashion way (email).

**Comment by Unknwon** [comment]
@Eriner upgrade security is good but I think it has not been a must thing here since PBKDF2 is relatively secure but still secure.

So may leave this change until 1.0 or some early releases like 0.8 so people can have some to upgrade.

**Comment by Eriner** [comment]
@Unknwon I agree. PBKDF2 is still secure, so the move to bcrypt would be nice, but not an immediate issue. It would be nice to have it in the 1.0 release, though, as that point is when code-cleanup and improvements like this should happen. Cheers.

**Comment by TuxCoder** [comment]
@Eriner  Thank you for your good description about the differences. 
Yeah I would also agree that PBKDF2 is still secure enough if you choose a not too bad password.
@Unknwon Would be nice to have bcrypt or scrypt in the final version.

Probably if I find time I will submit a pullrequest but I'm not sure when this will be.

**Comment by Eriner** [comment]
Thanks @TuxCoder. If you get to working on it, ping me and I'll take a look at it before you PR.

**Comment by Unknwon** [comment]
@TuxCoder Thanks, PR is good but just to be clear it may be merged far after you proposed. :smiley:

**Comment by gronke** [comment]
#3662 now comes with BCRYPT support 🎉  You can do me a favor and review that branch!
